examples/tutorials/matmul/int4_weight_decompression.cpp

```diff
@@ -0,0 +1,264 @@
+/// to choose the most appropriate layout for best performance.
+///
+/// @warning
+/// The format tag #dnnl::memory::format_tag::any doesn't work for memory
+/// descriptors that have one or more unknown dimensions and/or strides.
+///
+/// @include int4_weight_decompression.cpp
+#include <cassert>
+#include <cctype>
+#include <cmath>
+#include <cstdio>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <vector>
+
+#include "oneapi/dnnl/dnnl.hpp"
+
+#include "example_utils.hpp"
+
+using namespace dnnl;
+
+namespace {
+
+void init_vector(std::vector<float> &v) {
+    std::mt19937 gen;
+    std::uniform_real_distribution<float> u(0, 1);
+    for (auto &e : v)
+        e = u(gen);
+}
+
+void init_vector(std::vector<int32_t> &v) {
+    std::mt19937 gen;
+    std::uniform_int_distribution<int32_t> u(0, 511);
+    for (auto &e : v)
+        e = u(gen);
+}
+
+// Transpose the INT4 data for the src vector which packs 8 INT4 values as INT32,
+// for example, the data babababa is transposed to abababab.
+void transpose_s4(const std::vector<int32_t> &src, std::vector<int32_t> &dst,
+        int32_t K, int32_t N) {
+    // Ensure dst is the correct size
+    dst.resize(src.size());
+
+    // Iterate over the src vector to transpose the INT4 data
+    for (int32_t k = 0; k < K; ++k) {
+        for (int32_t n = 0; n < N; ++n) {
+            // Extract INT4 values from src
+            int32_t src_byte = src[k * N + n];
+            for (int32_t i = 0; i < 8; ++i) {
+                int8_t src_int4 = (src_byte >> (i * 4)) & 0x0F;
+
+                // Calculate destination indices
+                int32_t dst_index = (n * K + k) / 8;
+                int32_t dst_offset = (n * K + k) % 8;
+
+                // Pack INT4 values into dst
+                dst[dst_index] &= ~(
+                        0x0F << (dst_offset * 4)); // Clear the destination INT4
+                dst[dst_index] |= (src_int4
+                        << (dst_offset * 4)); // Set the destination INT4
+            }
+        }
+    }
+}
+
+} // namespace
+
+int32_t number_of_runs = 1;
+
+// Create a MatMul primitive descriptor for the following op:
+// C_f16 = A_f16 * (B_s4 - zp_B) * sc_B[:]
+//
+// Here:
+// - Matrices A and C are of f16 data type.
+// - The B matrix is stored as s4 with format tag ab, its zero point is zp_B,
+//   and all its dimensions are known. This matrix can be a matrix of compressed
+//   weights in an MLP topology.
+// - The weights scaling and zero point values are not known at the primitive creation time.
+matmul::primitive_desc matmul_pd_create(int64_t M, int64_t N, int64_t K,
+        int64_t G_SC, int64_t G_ZP, const engine &eng) {
+
+    memory::desc a_md({M, K}, memory::data_type::f16,
+            memory::format_tag::ab); // M x K layout
+    // oneDNN doesn't have a notion of format for zero-points and it's always considered as tag::ab
+    // In this example, we align the weights format to match the format tag::ab of the zero-points
```

Comment:
IIRC, PyTorch already had weights in `ba` which suits computations better. And zero-points and scales were in `ba` as well which was the problem for them. It's probably better to show how to transpose zero-points (which I can see below in the code) and scales instead because it's less memory to touch and should be more effective at the end.